_posts/2023-01-25-auto-restarting-connectors.md

```diff
@@ -0,0 +1,317 @@
+    value.converter: org.apache.kafka.connect.storage.StringConverter
+    key.converter.schemas.enable: false
+    value.converter.schemas.enable: false
+```
+
+After we create the connector, it creates the task and runs without any issues because it is not receiving any messages yet.
+So we have to start a producer and send the first 5 messages.
+
+```
+kubectl run kafka-producer -ti --image=quay.io/strimzi/kafka:0.33.0-kafka-3.3.2 --rm=true --restart=Never -- bin/kafka-console-producer.sh --bootstrap-server my-cluster-kafka-bootstrap:9092 --topic my-topic
+If you don't see a command prompt, try pressing enter.
+>Hello World 1
+>Hello World 2
+>Hello World 3
+>Hello World 4
+>Hello World 5
+```
+
+After the fifth message, the following error shows in the Kafka Connect log:
+
+```
+2023-01-22 23:02:52,246 WARN [echo-sink|task-0] Failing as requested after 5 records (cz.scholz.kafka.connect.echosink.EchoSinkTask) [task-thread-echo-sink-0]
+2023-01-22 23:02:52,246 ERROR [echo-sink|task-0] WorkerSinkTask{id=echo-sink-0} Task threw an uncaught and unrecoverable exception. Task is being killed and will not recover until manually restarted. Error: Intentional task failure after receiving 5 records. (org.apache.kafka.connect.runtime.WorkerSinkTask) [task-thread-echo-sink-0]
+java.lang.RuntimeException: Intentional task failure after receiving 5 records.
+	at cz.scholz.kafka.connect.echosink.EchoSinkTask.put(EchoSinkTask.java:131)
+	at org.apache.kafka.connect.runtime.WorkerSinkTask.deliverMessages(WorkerSinkTask.java:581)
+	at org.apache.kafka.connect.runtime.WorkerSinkTask.poll(WorkerSinkTask.java:333)
+	at org.apache.kafka.connect.runtime.WorkerSinkTask.iteration(WorkerSinkTask.java:234)
+	at org.apache.kafka.connect.runtime.WorkerSinkTask.execute(WorkerSinkTask.java:203)
+	at org.apache.kafka.connect.runtime.WorkerTask.doRun(WorkerTask.java:189)
+	at org.apache.kafka.connect.runtime.WorkerTask.run(WorkerTask.java:244)
+	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
+	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
+	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
+	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
+	at java.base/java.lang.Thread.run(Thread.java:833)
+```
+
+And the task fails.
+In the next reconciliation, the operator restarts the task.
+You can see this in the operator log:
+
+```
+2023-01-22 23:04:24 INFO  AbstractOperator:239 - Reconciliation #73(timer) KafkaConnect(myproject/my-connect): KafkaConnect my-connect will be checked for creation or modification
+2023-01-22 23:04:24 INFO  AbstractConnectOperator:502 - Reconciliation #73(timer) KafkaConnect(myproject/my-connect): creating/updating connector: echo-sink
+2023-01-22 23:04:24 INFO  AbstractConnectOperator:692 - Reconciliation #73(timer) KafkaConnect(myproject/my-connect): Auto restarting connector echo-sink
```

Comment:
Not about this PR but I was wondering if we could make the log clear that it's autorestarting because of a failure and if it is possible to get more info about the failure from the Kafka Connect API.

Reply:
I guess that could be a useful enhancement. There is just one reason for auto-restarting - because of failure. So that is IMHO clear. But the error might be useful as at the info level it is currently not logged IIRC.